File: frontend/src/services/cli-bridge.test.ts
// @vitest-environment node
// ===== CLI Bridge 前端封装层测试 =====
// 测试 cli-bridge.ts 的命令执行、响应解析、网页版降级等功能

import { describe, it, expect, beforeEach, vi } from "vitest";
import {
  executeCLI,
  getAllowedCLICommands,
  cliSearch,
  cliList,
  cliAnalyze,
  cliCacheStatus,
  ALLOWED_CLI_COMMANDS,
  parseCLIResponse,
  buildArgsMap,
} from "./cli-bridge.ts";

// Mock getApp 和 resolveWebMode
vi.mock("../backend/app.ts", () => ({
  getApp: vi.fn(),
}));

vi.mock("../backend/platform.ts", () => ({
  resolveWebMode: vi.fn(() => false),
}));

vi.mock("../backend/web-common.ts", () => ({
  WebUnsupportedError: class extends Error {
    constructor(binding: string) {
      super(`[web] binding ${binding} 浏览器端未实现`);
      this.name = "WebUnsupportedError";
    }
  },
}));

import { getApp } from "../backend/app.ts";
import { resolveWebMode } from "../backend/platform.ts";
import { WebUnsupportedError } from "../backend/web-common.ts";

// 重置 mock
beforeEach(() => {
  vi.clearAllMocks();
  vi.mocked(resolveWebMode).mockReturnValue(false);
});

describe("CLI Bridge - 命令执行", () => {
  it("执行白名单命令成功", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({
          status: "success",
          command: "search",
          data: { output: "test output", lines: ["line1"] },
          timing: { total_ms: 100 },
          meta: { platform: "windows" },
        })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    const result = await executeCLI("search", { keyword: "test" });

    expect(result.status).toBe("success");
    expect(result.command).toBe("search");
    expect(result.data?.output).toBe("test output");
    expect(result.timing?.total_ms).toBe(100);
  });

  it("拒绝非白名单命令", async () => {
    const result = await executeCLI("dangerous-command");

    expect(result.status).toBe("not_supported");
    expect(result.error?.code).toBe("command_not_allowed");
    expect(result.command).toBe("dangerous-command");
  });

  it("处理执行错误", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({
          status: "error",
          command: "search",
          error: { code: "runtime_error", message: "模型不存在" },
          timing: { total_ms: 50 },
        })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    const result = await executeCLI("search", {});

    expect(result.status).toBe("error");
    expect(result.error?.code).toBe("runtime_error");
    expect(result.error?.message).toBe("模型不存在");
  });

  it("处理 getApp 异常", async () => {
    vi.mocked(getApp).mockRejectedValue(new Error("连接失败"));

    const result = await executeCLI("search");

    expect(result.status).toBe("error");
    expect(result.error?.code).toBe("call_failed");
  });
});

describe("CLI Bridge - 网页版降级", () => {
  it("网页版返回降级响应", async () => {
    vi.mocked(resolveWebMode).mockReturnValue(true);

    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({
          status: "not_supported",
          command: "cache-status",
          error: {
            code: "web_not_supported",
            message: "网页版不支持命令 [cache-status]",
          },
          meta: { platform: "web" },
        })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    const result = await executeCLI("cache-status");

    expect(result.status).toBe("not_supported");
    expect(result.error?.code).toBe("web_not_supported");
    expect(result.meta?.platform).toBe("web");
  });

  it("网页版 WebUnsupportedError 处理", async () => {
    vi.mocked(resolveWebMode).mockReturnValue(true);

    const mockApp = {
      ExecuteCLI: vi.fn().mockImplementation(() => {
        throw new WebUnsupportedError("ExecuteCLI");
      }),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    const result = await executeCLI("analyze");

    expect(result.status).toBe("not_supported");
    expect(result.error?.code).toBe("web_not_supported");
  });
});

describe("CLI Bridge - 命令列表", () => {
  it("获取允许的命令列表（桌面版）", async () => {
    const commands = ["search", "list", "analyze"];
    const mockApp = {
      GetAllowedCLICommands: vi.fn().mockResolvedValue(JSON.stringify(commands)),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    const result = await getAllowedCLICommands();

    expect(result).toEqual(commands);
    expect(mockApp.GetAllowedCLICommands).toHaveBeenCalled();
  });

  it("获取允许的命令列表（网页版降级）", async () => {
    vi.mocked(resolveWebMode).mockReturnValue(true);

    const result = await getAllowedCLICommands();

    expect(result).toEqual(ALLOWED_CLI_COMMANDS);
  });

  it("getApp 异常时降级到本地列表", async () => {
    vi.mocked(getApp).mockRejectedValue(new Error("连接失败"));

    const result = await getAllowedCLICommands();

    expect(result).toEqual(ALLOWED_CLI_COMMANDS);
  });
});

describe("CLI Bridge - 便捷方法", () => {
  it("cliSearch 传递正确参数", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({ status: "success", command: "search", data: {} })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    await cliSearch({ keyword: "warrior", format: "json" });

    expect(mockApp.ExecuteCLI).toHaveBeenCalledWith("search", {
      keyword: "warrior",
      format: "json",
    });
  });

  it("cliList 调用 list 命令", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({ status: "success", command: "list", data: {} })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    await cliList();

    expect(mockApp.ExecuteCLI).toHaveBeenCalledWith("list", {});
  });

  it("cliAnalyze 传递 model 参数", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({ status: "success", command: "analyze", data: {} })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    await cliAnalyze({ model: "/path/to/model.ysm" });

    expect(mockApp.ExecuteCLI).toHaveBeenCalledWith("analyze", {
      model: "/path/to/model.ysm",
    });
  });

  it("cliCacheStatus 调用 cache-status 命令", async () => {
    const mockApp = {
      ExecuteCLI: vi.fn().mockResolvedValue(
        JSON.stringify({ status: "success", command: "cache-status", data: {} })
      ),
    };

    vi.mocked(getApp).mockResolvedValue(mockApp);

    await cliCacheStatus();

    expect(mockApp.ExecuteCLI).toHaveBeenCalledWith("cache-status", {});
  });
});

describe("CLI Bridge - 响应解析", () => {
  it("解析有效 JSON 响应", () => {
    const raw = JSON.stringify({
      status: "success",
      command: "test",
      data: { key: "value" },
    });

    const result = parseCLIResponse(raw);

    expect(result.status).toBe("success");
    expect(result.command).toBe("test");
    expect(result.data?.key).toBe("value");
  });

  it("解析无效 JSON 返回错误", () => {
    const result = parseCLIResponse("invalid json");

    expect(result.status).toBe("error");
    expect(result.error?.code).toBe("parse_error");
  });
});

describe("CLI Bridge - 参数构建", () => {
  it("过滤 undefined 值", () => {
    const result = buildArgsMap({
      keyword: "test",
      format: undefined,
      count: 5,
      verbose: true,
    });

    expect(result).toEqual({
      keyword: "test",
      count: 5,
      verbose: true,
    });
    expect(result).not.toHaveProperty("format");
  });

  it("过滤 null 值", () => {
    const result = buildArgsMap({
      keyword: null,
      valid: "yes",
    });

    expect(result).toEqual({ valid: "yes" });
    expect(result).not.toHaveProperty("keyword");
  });

  it("空对象返回空 map", () => {
    const result = buildArgsMap({});

    expect(result).toEqual({});
  });
});

describe("CLI Bridge - 白名单常量", () => {
  it("包含所有预期命令", () => {
    const expected = [
      "search", "analyze", "list", "verify", "benchmark", "export",
      "file-bench", "single-bench", "concurrent-bench",
      "scan-dir", "analyze-mmd", "perf-log",
      "cache-status", "cache-verify", "cache-clear", "cache-diag",
      "config-show", "gui-flow",
    ];

    for (const cmd of expected) {
      expect(ALLOWED_CLI_COMMANDS).toContain(cmd);
    }
  });

  it("数量正确", () => {
    expect(ALLOWED_CLI_COMMANDS).toHaveLength(18);
  });

  it("类型安全：只能传入白名单中的命令", () => {
    // 验证 AllowedCLICommand 类型
    const validCmd: (typeof ALLOWED_CLI_COMMANDS)[number] = "search";
    expect(ALLOWED_CLI_COMMANDS).toContain(validCmd);
  });
});

File: internal/app/cli_bridge_test.go
package app

import (
	"encoding/json"
	"strings"
	"testing"
)

// TestExecuteCLI_CommandNotAllowed 测试命令不在白名单中
func TestExecuteCLI_CommandNotAllowed(t *testing.T) {
	a := NewApp()
	result := a.ExecuteCLI("unknown-command", nil)

	var resp map[string]interface{}
	if err := json.Unmarshal([]byte(result), &resp); err != nil {
		t.Fatalf("JSON 解析失败: %v", err)
	}

	if resp["status"] != "not_supported" {
		t.Errorf("期望 status=not_supported, 实际=%v", resp["status"])
	}

	if resp["command"] != "unknown-command" {
		t.Errorf("期望 command=unknown-command, 实际=%v", resp["command"])
	}
}

// TestExecuteCLI_CommandAllowed 测试白名单中的命令（不需要真实文件系统）
func TestExecuteCLI_CommandAllowed(t *testing.T) {
	a := NewApp()

	// 测试 config-show 命令（不需要文件系统操作）
	result := a.ExecuteCLI("config-show", map[string]interface{}{})

	var resp map[string]interface{}
	if err := json.Unmarshal([]byte(result), &resp); err != nil {
		t.Fatalf("JSON 解析失败: %v", err)
	}

	// 应该有 status 字段
	if _, ok := resp["status"]; !ok {
		t.Error("响应缺少 status 字段")
	}

	// 应该有 command 字段
	if resp["command"] != "config-show" {
		t.Errorf("期望 command=config-show, 实际=%v", resp["command"])
	}
}

// TestExecuteCLI_AllAllowedCommands 测试所有白名单命令都能通过验证
func TestExecuteCLI_AllAllowedCommands(t *testing.T) {
	a := NewApp()
	commands := []string{
		"search", "analyze", "list", "verify", "benchmark", "export",
		"file-bench", "single-bench", "concurrent-bench",
		"scan-dir", "analyze-mmd", "perf-log",
		"cache-status", "cache-verify", "cache-clear", "cache-diag",
		"config-show", "gui-flow",
	}

	for _, cmd := range commands {
		result := a.ExecuteCLI(cmd, map[string]interface{}{})
		var resp map[string]interface{}
		if err := json.Unmarshal([]byte(result), &resp); err != nil {
			t.Errorf("命令 %s JSON 解析失败: %v", cmd, err)
			continue
		}
		// 应该返回 success 或 error（不是 not_supported）
		status := resp["status"]
		if status == "not_supported" {
			t.Errorf("命令 %s 不应返回 not_supported", cmd)
		}
	}
}

// TestExecuteCLI_ArgsBuilding 测试参数构建逻辑
func TestExecuteCLI_ArgsBuilding(t *testing.T) {
	a := NewApp()

	// 测试带参数的命令
	result := a.ExecuteCLI("search", map[string]interface{}{
		"keyword": "test",
		"format":  "json",
	})

	var resp map[string]interface{}
	if err := json.Unmarshal([]byte(result), &resp); err != nil {
		t.Fatalf("JSON 解析失败: %v", err)
	}

	// 验证响应结构
	if _, ok := resp["timing"]; !ok {
		t.Error("响应缺少 timing 字段")
	}

	if _, ok := resp["meta"]; !ok {
		t.Error("响应缺少 meta 字段")
	}
}

// TestGetAllowedCLICommands 测试获取允许的命令列表
func TestGetAllowedCLICommands(t *testing.T) {
	a := NewApp()
	result := a.GetAllowedCLICommands()

	var commands []string
	if err := json.Unmarshal([]byte(result), &commands); err != nil {
		t.Fatalf("JSON 解析失败: %v", err)
	}

	expectedCount := 20 // 与 allowedCLICommands 数量一致
	if len(commands) != expectedCount {
		t.Errorf("期望 %d 个命令, 实际 %d 个: %v", expectedCount, len(commands), commands)
	}

	// 检查关键命令是否存在
	expectedCmds := []string{"search", "list", "analyze", "config-show"}
	for _, cmd := range expectedCmds {
		found := false
		for _, c := range commands {
			if c == cmd {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("命令 %s 不在列表中", cmd)
		}
	}
}

// TestMakeJsonResponse 测试 JSON 响应构建
func TestMakeJsonResponse(t *testing.T) {
	t.Run("success response", func(t *testing.T) {
		result := makeJsonResponse("success", "test-cmd", map[string]string{"key": "value"}, nil, 100.0)

		var resp map[string]interface{}
		if err := json.Unmarshal([]byte(result), &resp); err != nil {
			t.Fatalf("JSON 解析失败: %v", err)
		}

		if resp["status"] != "success" {
			t.Errorf("期望 status=success, 实际=%v", resp["status"])
		}
		if resp["command"] != "test-cmd" {
			t.Errorf("期望 command=test-cmd, 实际=%v", resp["command"])
		}

		// data 应该存在
		if resp["data"] == nil {
			t.Error("data 不应为 nil")
		}
	})

	t.Run("error response", func(t *testing.T) {
		errData := map[string]string{"code": "test_error", "message": "test error message"}
		result := makeJsonResponse("error", "test-cmd", nil, errData, 50.0)

		var resp map[string]interface{}
		if err := json.Unmarshal([]byte(result), &resp); err != nil {
			t.Fatalf("JSON 解析失败: %v", err)
		}

		if resp["status"] != "error" {
			t.Errorf("期望 status=error, 实际=%v", resp["status"])
		}
		if resp["data"] != nil {
			t.Error("data 应为 nil")
		}

		// error 应该存在
		if resp["error"] == nil {
			t.Error("error 不应为 nil")
		}
	})

	t.Run("timing info", func(t *testing.T) {
		result := makeJsonResponse("success", "cmd", nil, nil, 123.456)

		var resp map[string]interface{}
		if err := json.Unmarshal([]byte(result), &resp); err != nil {
			t.Fatalf("JSON 解析失败: %v", err)
		}

		timing, ok := resp["timing"].(map[string]interface{})
		if !ok {
			t.Fatal("timing 字段格式错误")
		}

		totalMs, ok := timing["total_ms"].(float64)
		if !ok {
			t.Fatal("timing.total_ms 字段类型错误")
		}
		if totalMs != 123.456 {
			t.Errorf("期望 total_ms=123.456, 实际=%v", totalMs)
		}
	})

	t.Run("meta info", func(t *testing.T) {
		result := makeJsonResponse("success", "cmd", nil, nil, 0.0)

		var resp map[string]interface{}
		if err := json.Unmarshal([]byte(result), &resp); err != nil {
			t.Fatalf("JSON 解析失败: %v", err)
		}

		meta, ok := resp["meta"].(map[string]interface{})
		if !ok {
			t.Fatal("meta 字段格式错误")
		}

		platform, ok := meta["platform"].(string)
		if !ok {
			t.Fatal("meta.platform 字段类型错误")
		}
		if platform == "" {
			t.Error("meta.platform 不应为空")
		}
	})
}

// TestSplitLines 测试字符串分行
func TestSplitLines(t *testing.T) {
	tests := []struct {
		input    string
		expected []string
	}{
		{"", []string{}},
		{"single line", []string{"single line"}},
		{"line1\nline2", []string{"line1", "line2"}},
		{"line1\nline2\nline3", []string{"line1", "line2", "line3"}},
		{"\nline2", []string{"line2"}},
		{"line1\n", []string{"line1"}},
	}

	for _, tt := range tests {
		result := splitLines(tt.input)
		if len(result) != len(tt.expected) {
			t.Errorf("splitLines(%q): 期望 %d 行, 实际 %d 行: %v", tt.input, len(tt.expected), len(result), result)
			continue
		}
		for i, line := range result {
			if line != tt.expected[i] {
				t.Errorf("splitLines(%q)[%d]: 期望 %q, 实际 %q", tt.input, i, tt.expected[i], line)
			}
		}
	}
}

// TestAllowedCommandsCount 测试白名单命令数量与 json.go 保持一致
func TestAllowedCommandsCount(t *testing.T) {
	// cli_bridge.go 中的白名单命令
	bridgeCount := len(allowedCLICommands)

	// json.go 中的白名单命令（从文档/注释推断）
	// 两者应保持同步
	expectedCommands := []string{
		"search", "analyze", "list", "verify", "benchmark", "export",
		"file-bench", "single-bench", "concurrent-bench",
		"scan-dir", "analyze-mmd", "perf-log",
		"cache-status", "cache-verify", "cache-clear", "cache-diag",
		"config-show", "gui-flow",
		"resource-scan", "repo-audit",
	}

	if bridgeCount != len(expectedCommands) {
		t.Errorf("白名单命令数量不匹配: cli_bridge.go=%d, 期望=%d", bridgeCount, len(expectedCommands))
	}

	// 检查所有期望的命令都在白名单中
	for _, cmd := range expectedCommands {
		if !allowedCLICommands[cmd] {
			t.Errorf("命令 %s 不在白名单中", cmd)
		}
	}
}

// TestExecuteCLI_InvalidJSONResponse 测试参数 map 中不同类型值的处理
func TestExecuteCLI_InvalidJSONResponse(t *testing.T) {
	a := NewApp()

	// 测试字符串参数
	result := a.ExecuteCLI("search", map[string]interface{}{
		"keyword": "test keyword",
	})
	if !strings.Contains(result, `"status"`) {
		t.Error("响应应包含 status 字段")
	}

	// 测试数字参数（float64）
	result = a.ExecuteCLI("benchmark", map[string]interface{}{
		"iterations": float64(5),
	})
	if !strings.Contains(result, `"status"`) {
		t.Error("响应应包含 status 字段")
	}

	// 测试布尔参数
	result = a.ExecuteCLI("verify", map[string]interface{}{
		"repair": true,
	})
	if !strings.Contains(result, `"status"`) {
		t.Error("响应应包含 status 字段")
	}
}
